src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1716,38 +1807,47 @@
-  ;; drop partition tables, and row exclusive to update certnames.
-  (acquire-locks! {"certnames" "ROW EXCLUSIVE"
-                   "reports" "ACCESS EXCLUSIVE"
-                   "resource_events" "ACCESS EXCLUSIVE"})
-  ;; force a resource-events GC. prior to partitioning, this would have happened
-  ;; via a cascade when the report was deleted, but now we just drop whole tables
-  ;; of resource events.
-  (delete-resource-events-older-than! (if (before? report-ttl resource-events-ttl)
-                                        resource-events-ttl report-ttl)
-                                      incremental?
-                                      update-lock-status)
-  (prune-daily-partitions "reports" report-ttl incremental?
-                          update-lock-status :write-locking-reports)
-  ;; since we cannot cascade back to the certnames table anymore, go clean up
-  ;; the latest_report_id column after a GC
-  (jdbc/do-commands
-   ["UPDATE certnames SET latest_report_id = NULL"
-    "  WHERE certname IN"
-    "    (SELECT DISTINCT certnames.certname FROM certnames"
-    "       LEFT OUTER JOIN reports ON (reports.certname = certnames.certname)"
-    "       WHERE reports.id IS NULL)"]))
-
-
+  (let [effective-resource-events-ttl
+         (if (before? report-ttl resource-events-ttl) resource-events-ttl report-ttl)]
+    (if-not (detach-partitions-concurrently?)
+      ;; PG11
+      (delete-reports-older-than-in-pg-11! report-ttl effective-resource-events-ttl
+                                           incremental? update-lock-status)
+      ;; PG14+
+      ;; Detach partition concurrently must take place outside of a transaction.
+      (let [detached-resource-event-tables
+             (detach-daily-partitions "resource_events" effective-resource-events-ttl
+                                      incremental? update-lock-status
+                                      :write-locking-resource-events)
+            detached-report-tables
+             (detach-daily-partitions "reports" report-ttl
+                                    incremental? update-lock-status
+                                    :write-locking-reports)]
+        ;; Now we can delete the partitions with less intrusive locking.
+        (jdbc/with-db-transaction []
+          ;; We update these tables in a different order than report insertion, creates
+          ;; the potential for deadlocks when two concurrent transactions have each
+          ;; acquired a portion of their locks. So first, we acquire locks on all the
+          ;; tables in the same order they are accessed during insertion. These match
+          ;; the locks that would implicitly be acquired by the operations we're going
+          ;; to perform: Access exclusive for reports and resource_events so we can
+          ;; drop partition tables, and row exclusive to update certnames.
+          (acquire-locks! {"certnames" "ROW EXCLUSIVE"
+                           "reports" "ACCESS EXCLUSIVE"
```

Comment:
I believe these locks have to remain access exclusive in order to drop the tables safely. So whatever benefit we are seeing is presumably because the detach means they should no longer be being queried?

Reply:
Adjusted this in 133c377